ElunaTemplate.h

```diff
@@ -461,13 +417,38 @@ class ElunaTemplate
     {
         ElunaRegister<T>* l = static_cast<ElunaRegister<T>*>(lua_touserdata(L, lua_upvalueindex(1)));
         Eluna* E = static_cast<Eluna*>(lua_touserdata(L, lua_upvalueindex(2)));
+        T* obj;
 
-        T* obj = E->CHECKOBJ<T>(1); // get self
-        if (!obj)
-            return 0;
+        // determine if the method table functions are global or non-global
+        bool isGlobal = false;
+        std::visit([&](auto&& func)
+        {
+            using FuncType = std::decay_t<decltype(func)>;
+            if constexpr (std::is_same_v<FuncType, int(*)(Eluna*)>)
+                isGlobal = true;
+        }, l->mfunc);
+
+        // we only check self if the method is not a global
+        if (!isGlobal)
+        {
+            obj = E->CHECKOBJ<T>(1);
+            if (!obj)
+                return 0;
+        }
 
         int top = lua_gettop(L);
-        int expected = l->mfunc(E, obj);
+
+        int expected = 0;
+        std::visit([&](auto&& func)
+        {
+            using FuncType = std::decay_t<decltype(func)>;
+            if constexpr (std::is_same_v<FuncType, int(*)(Eluna*)>) // global method
+                expected = func(E);
+            else if constexpr (std::is_same_v<FuncType, int(*)(Eluna*, T*)>) // non-global method
+                expected = func(E, obj);
+
+        }, l->mfunc);
```

Comment:
I think we can just do this so its simpler and maybe more efficient also

```cpp
        // determine if the method table functions are global or non-global
        constexpr bool isGlobal = std::is_same_v<T, void>;

        // we only check self if the method is not a global
        if constexpr (!isGlobal)
        {
            /// code ....
        }

        int top = lua_gettop(L);

        int expected = 0;
        if constexpr (!isGlobal) {
            expected = func(E);
        } else {
            expected = func(E, obj);
        }
```

Reply:
I tried something similar, but the compiler really doesn't like the func calls that way, they do not evaluate to a function taking either one or two arguments. It also complains about visit not matching any overloaded functions.

Reply:
Pushed a new version of thunk that doesn't use lambdas